### PR TITLE
Combine lib and device into a single backend

### PR DESCRIPTION
**Pull request by vmalia**
# Description

Improve compile-time dispatching mechanism to make it more flexible for enabling new backends, devices, and domains.
* Change two-parameter templates into one-parameter template (<library, backend> to <backend>)
* Change the backend naming convention from device name to library/device name for oneMKL (intelcpu -> mklcpu). cublas is still cublas.
* Updated headers, scripts, test_helper and docs.

# Checklist
- [x] Do all unit tests pass locally? Yes.
- [x] Have you formatted the code using clang-format? Yes

### CODE REVIEW

File: include/oneapi/mkl/blas/detail/blas_ct_templates.hpp [hunk 1]
```diff
@@ -1,347 +1,328 @@
-/*******************************************************************************
-* Copyright 2020 Intel Corporation
```
Comment (mkrainiuk) [inline include/oneapi/mkl/blas/detail/blas_ct_templates.hpp:2]: We need Intel copyrights here

File: include/oneapi/mkl/blas/detail/cublas/blas_ct.hpp [hunk 2]
```diff
@@ -16,717 +16,673 @@
 *  limitations under the License.
 *
 **************************************************************************/
+
 //
-// Generated based on onemkl/blas/blas.hpp
+// Generated based on include/oneapi/mkl/blas/detail/blas_ct_templates.hpp
 //
 
-#ifndef _DETAIL_CUBLAS_BLAS_HPP_
```
Comment (mkrainiuk) [inline include/oneapi/mkl/blas/detail/cublas/blas_ct.hpp:23]: I believe in other header files we are using macro guards, does it make sense to keep it aligned with them?

File: include/oneapi/mkl/blas/detail/mklgpu/blas_ct.hpp [hunk 3]
```diff
@@ -1,735 +1,670 @@
-/*******************************************************************************
-* Copyright 2020 Intel Corporation
```
Comment (mkrainiuk) [inline include/oneapi/mkl/blas/detail/mklgpu/blas_ct.hpp:2]: Need Intel copyright

File: include/oneapi/mkl/blas/detail/blas_ct_templates.hpp (hunk 1 above)
Comment (vmalia) [reply include/oneapi/mkl/blas/detail/blas_ct_templates.hpp]: added

File: include/oneapi/mkl/blas/detail/cublas/blas_ct.hpp [hunk 4]
```diff
@@ -27,706 +28,661 @@
 #include <complex>
 #include <cstdint>
 
-#include "oneapi/mkl/detail/backends.hpp"
-#include "oneapi/mkl/detail/libraries.hpp"
 #include "oneapi/mkl/types.hpp"
-
-#include "onemkl_blas_cublas.hpp"
```
Comment (mkrainiuk) [inline include/oneapi/mkl/blas/detail/cublas/blas_ct.hpp:34]: This header file has cublas backend functions declaration, I believe we need it to be able to call oneapi::mkl::cublas functions from the wrappers below.

Comment (vmalia) [reply include/oneapi/mkl/blas/detail/cublas/blas_ct.hpp:34]: added it back

**Pull request merged**